### ISSUE TITLE
Multiple websocket resources

### ISSUE THREAD

**Issue by Seiguerboujf** [warmcat/libwebsockets#2277]
I had some questions regarding Libwebsockets. I created a server that has 5 Http resources "mountpoints" and 1 websocket resource. So far my program is working and I am able to test it using Postman and websocat.
I want to add another websocket resource so that in total I get 2 websocket resources and I couldn't so my question is, is that possible using libwebsockets?

For some reason, I need to put the protocol handler for the websocket resource as the first element of the struct lws_protocols array in order for it to work.(this is a problem itself)

also when added another websocket callback (protocol) if i run both with websocat even on different ports websocat ws://127.0.0.1:7682/ev/ and websocat ws://127.0.0.1:7681/state/ one of them crashes.
I am not using a .js file or a html page to test my server just using postman for http and websocat from linux terminal

**Comment by lws-team** [comment]
You can have many ws protocols on your server OK.

But you cannot have more than one "default" protocol.  This is the protocol that is selected if you do not give a subprotocol name when opening the socket.  Actually, ws is designed that the JS opening the ws provides the name of the subprotocol they want on the server, it is the second argument like this in the JS `new WebSocket(urlpath, protocolname);`.  This adds a Sec-websocket-protocol header to the browser's request with the name, lws then selects which protocol to bind to by matching the name.

https://tools.ietf.org/html/rfc6455#section-1.9

Most examples on the net dumb it down and leave the protocolname out, that selects the "default" protocol on the server.  But of course you can only have one of those, and it defaults in lws to being the first protocol in the list.

So you can get rid of those issues by asking for the name of the ws protocol you want on the server in the JS, and making sure it matches the protocol name you are giving lws.

> one of them crashes.

websocat crashes?  Dunno.  lws crashes?  let's have a backtrace.

**Comment by Seiguerboujf** [comment]
lws dosen't crash but websocat does.
I get this 
websocat: WebSocketError: UTF-8 failure
websocat: error running
on the terminal when I run the 2 websocket resources even thought I have two vhosts and each one of them is mounted to a port
in this example 7681 and 7682 
The problem may be websocat https://github.com/vi/websocat because I'm using it to test my ws protocols instead of a JS file and new WebSocket(urlpath, protocolname) to match the protocolname.

**Comment by lws-team** [comment]
You should only send valid UTF-8 on ws TEXT mode.  If you want to send binary, use BINARY mode.

I'm not sure websocat "crashes", it just complains, doesn't it?  Those are very different things.

**Comment by Seiguerboujf** [comment]
int s = lws_write(pss->wsi, ((unsigned char *) pmsg->payload) +LWS_PRE, pmsg->len, LWS_WRITE_TEXT); 

This is what I'm sending on both websocket resources and the one on port 7862 shows me the message above when I connect the one on port 7681 and this one works fine it doesn't show any errors

**Comment by Seiguerboujf** [comment]
Yes it is.

**Comment by Seiguerboujf** [comment]
Is it possible to test one of the minimal ws server examples whithout using the JS code? or Is it necessary in order to "know" what protocol we're using?

**Comment by lws-team** [comment]
Well, either it is valid UTF-8 and websocat is telling you wrong, or you are not sending what you think you are sending.

You can pop a `lws_hexdump_notice(((unsigned char *) pmsg->payload) +LWS_PRE, pmsg->len);` before the write and check what you are actually sending.

> Is it possible to test one of the minimal ws server examples whithout using the JS code? or Is it necessary in order to "know" what protocol we're using?

you must talk to a ws server using ws protocol.  From JS or some other ws client.  If you want more than one ws protocol per server, just must send the subprotocol names.

**Comment by Seiguerboujf** [comment]
OK got it thank you so much
